### PR TITLE
API Changes and CO Timestamp support

### PR DESCRIPTION
**Pull request by pradeepsixer**
1. Updated the API to support customizations by the API consumer
2. Added support for Created and Closed Timestamps for Change Requests.

### CODE REVIEW

File: testapp/tests.py [hunk 1]
```diff
@@ -60,86 +62,187 @@ def test_create_change_request(self, mock_get_snow_group_guid, mock_pysnow):
             'assignment_group': {'value': uuid.uuid4()},
             'state': '2'
         }
-        self.mock_pysnow_client.insert.return_value = fake_insert_retval
+
+        fake_resource = mock.MagicMock()
+        fake_resource.create.return_value = fake_insert_retval
+
+        self.mock_pysnow_client.resource.return_value = fake_resource
         mock_pysnow.Client.return_value = self.mock_pysnow_client
```
Comment (tarkatronic) [inline testapp/tests.py:73]: Can you get rid of the `self.mock_pysnow_client`? Since it's just an empty mock, it's not extremely useful, and it's making this logic a bit hard to follow.

File: testapp/tests.py [hunk 2]
```diff
@@ -60,86 +62,187 @@ def test_create_change_request(self, mock_get_snow_group_guid, mock_pysnow):
...
 
-        co = self.change_request_handler.create_change_request(
-            'title', 'description', assignment_group='assignment_group'
-        )
+        co = self.change_request_handler.create_change_request('Title', 'Description', payload={})
         last_co = ChangeRequest.objects.last()
 
-        mock_get_snow_group_guid.assert_called_with('assignment_group')
         self.assertEqual(co.pk, last_co.pk)
```
Comment (tarkatronic) [inline testapp/tests.py:78]: It appears that two separate behaviors are being tested here; one to make sure that the return value of the create is as expected, and one to ensure that the object is added to the db. If that is the case, these should probably be split into separate tests.

File: testapp/tests.py [hunk 3]
```diff
@@ -60,86 +62,187 @@ def test_create_change_request(self, mock_get_snow_group_guid, mock_pysnow):
...
         self.assertEqual(co.assignment_group_guid, fake_insert_retval['assignment_group']['value'])
 
-    @mock.patch('django_snow.helpers.snow_request_handler.ChangeRequestHandler._get_snow_group_guid')
-    def test_create_change_request_raises_exception_when_error_in_result(self, mock_get_snow_group_guid, mock_pysnow):
+    def test_create_change_request_parameters(self, mock_pysnow):
+        expected_payload = OrderedDict()
+        expected_payload['type'] = 'normal'
+        expected_payload['assignment_group'] = 'bar'
+        expected_payload['short_description'] = 'Title'
+        expected_payload['description'] = 'Description'
```
Comment (tarkatronic) [inline testapp/tests.py]: Is an `OrderedDict` actually necessary here? I'm guessing that `assert_called_with` cares about the order?

If so, this could be cleaned up a bit:
```python
expected_payload = OrderedDict({
    'type': 'normal',
    'assignment_group': 'bar',
    'short_description': 'Title',
    'description': 'Description'
})
```
The same applies below for `payload`.

File: testapp/tests.py [hunk 4]
```diff
@@ -60,86 +62,187 @@ def test_create_change_request(self, mock_get_snow_group_guid, mock_pysnow):
...
+        payload['assignment_group'] = 'bar'
+        self.change_request_handler.create_change_request('Title', 'Description', None, payload=payload)
+        fake_resource.create.assert_called_with(payload=expected_payload)
+
+    def test_create_change_request_default_parameters(self, mock_pysnow):
+        expected_payload = OrderedDict()
+        expected_payload['short_description'] = 'Title'
+        expected_payload['description'] = 'Description'
+        expected_payload['type'] = 'standard'
+        expected_payload['assignment_group'] = 'bar'
```
Comment (tarkatronic) [inline testapp/tests.py]: Same notes apply here about the `OrderedDict`.

File: testapp/tests.py [hunk 5]
```diff
@@ -60,86 +62,187 @@ def test_create_change_request(self, mock_get_snow_group_guid, mock_pysnow):
...
+
+        self.change_request_handler.group_guid_dict['assignment_group'] = 'bar'
```
Comment (tarkatronic) [inline testapp/tests.py]: I'm not particularly a fan of this. This seems like a bit of a TDD anti-pattern. With TDD, tests should be written before code, but this line is quite obviously written after the code, because it implies, and requires, an intimate knowledge of the code internals. Is there a different way this can be controlled? Perhaps using `django.test.override_settings`?

File: testapp/tests.py [hunk 6]
```diff
@@ -60,86 +62,187 @@ def test_create_change_request(self, mock_get_snow_group_guid, mock_pysnow):
...
         }
-        self.mock_pysnow_client.insert.return_value = fake_insert_retval
+
+        fake_resource = mock.MagicMock()
+        fake_resource.create.return_value = fake_insert_retval
+
+        self.mock_pysnow_client.resource.return_value = fake_resource
         mock_pysnow.Client.return_value = self.mock_pysnow_client
 
         with self.assertRaises(ChangeRequestException):
```
Comment (tarkatronic) [inline testapp/tests.py]: Since this is checking for the exact same exception as above, generated by the same method call, I think this should also verify that the proper error message is being raised, with `six.assertRaisesRegex()`.

File: testapp/tests.py [hunk 7]
```diff
@@ -60,86 +62,187 @@ def test_create_change_request(self, mock_get_snow_group_guid, mock_pysnow):
...
 
-    def test__get_snow_group_guid(self, mock_pysnow):
-        fake_query = mock.MagicMock()
-        fake_query.get_one.return_value = {'sys_id': 'yo'}
-        self.mock_pysnow_client.query.return_value = fake_query
+        fake_resource.update.return_value = {'error': '3'}
+        self.mock_pysnow_client.resource.return_value = fake_resource
+        mock_pysnow.Client.return_value = self.mock_pysnow_client
+
+        with self.assertRaises(ChangeRequestException):
```
Comment (tarkatronic) [inline testapp/tests.py]: The same notes about the assert apply here.

File: testapp/tests.py [hunk 8]
```diff
@@ -60,86 +62,187 @@ def test_create_change_request(self, mock_get_snow_group_guid, mock_pysnow):
...
+            self.change_request_handler.update_change_request(fake_change_order, payload='{"foo": "bar"}')
+
+    def test_get_snow_group_guid_cached_result(self, mock_pysnow):
+        self.change_request_handler.group_guid_dict['foo'] = 'bar'
+        self.assertEqual(self.change_request_handler.get_snow_group_guid('foo'), 'bar')
```
Comment (tarkatronic) [inline testapp/tests.py]: This test has the same code smell as I mentioned above with needing to have intimate knowledge of the inner workings of the code. I think it would be better to mock the pysnow result, and ensure that pysnow is only called on the first attempt to retrieve the guid. This same technique should probably be used in other tests. In fact, this could even justify the continued use of `self.mock_pysnow_client`, since that would be functionality which would be necessary for many of the tests.

**Review by tarkatronic — CHANGES_REQUESTED** [review]
I'm only reviewing the tests for a first pass, so that I can be sure they all make sense. To that end, I've made several comments about potential improvements.

One other thought I had is, does it perhaps make sense to split these tests out into multiple `TestCase` classes? That way there could be slightly better logical grouping of them, and more focused `setUp()` methods for each. Just a thought.

File: testapp/tests.py (hunk 2 above)
Comment (pradeepsixer) [reply testapp/tests.py:78]: With the line 71 gone, it focuses only on checking the values of the persistence object created.

**Comment by tarkatronic** [comment]
This PR fixes #2

File: testapp/tests.py (hunk 1 above)
Comment (pradeepsixer) [reply testapp/tests.py:73]: The `self.mock_pysnow_client` is being used in almost all the test cases. So instead of instantiating it every time within the test cases themselves, I think its better to have it at the setUp() method.

**Comment by tarkatronic** [comment]
The updates look great! Aside from one issue: `assertRaisesRegex` doesn't exist in Python 2.7; you will need to use the `six` module for this. That also means that you'll need to add it to the `deps` section in `tox.ini`.

**Review by tarkatronic — APPROVED** [review]
